src/app/encoder.cpp

```diff
@@ -0,0 +1,170 @@
+/**
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "chip-zcl-zpro-codec.h"
+#include <assert.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <support/BufBound.h>
+#include <support/logging/CHIPLogging.h>
+
+using namespace chip;
+extern "C" {
+
+static uint16_t doEncodeApsFrame(BufBound & buf, uint16_t profileID, uint16_t clusterId, uint8_t sourceEndpoint,
+                                 uint8_t destinationEndpoint, EmberApsOption options, uint16_t groupId, uint8_t sequence,
+                                 uint8_t radius)
+{
+
+    uint8_t control_byte = 0;
+    buf.Put(control_byte); // Put in a control byte
+    buf.PutLE16(profileID);
+    buf.PutLE16(clusterId);
+    buf.Put(sourceEndpoint);
+    buf.Put(destinationEndpoint);
+    buf.PutLE(options, sizeof(EmberApsOption));
+    buf.PutLE16(groupId);
+    buf.Put(sequence);
+    buf.Put(radius);
+
+    if (!buf.Fit() && buf.Size() > 0)
+    {
+        ChipLogError(Zcl, "Could not encode buffer due to insufficient space");
+    }
+    else
+    {
+        ChipLogProgress(Zcl, "Successfully encoded APS frame of size %d bytes", buf.Written());
+    }
+
+    return (buf.Fit() || buf.Size() == 0) ? buf.Written() : 0; // 0 size buffer implies measure case.
+}
+
+uint16_t encodeApsFrame(uint8_t * buffer, uint16_t buf_length, EmberApsFrame * apsFrame)
+{
+    if (buffer == nullptr)
+    {
+        buf_length = 0; // This is the measuring case.
+    }
+    BufBound buf = BufBound(buffer, buf_length);
+    return doEncodeApsFrame(buf, apsFrame->profileId, apsFrame->clusterId, apsFrame->sourceEndpoint, apsFrame->destinationEndpoint,
+                            apsFrame->options, apsFrame->groupId, apsFrame->sequence, apsFrame->radius);
+}
+
+uint32_t _encodeOnOffCommand(uint8_t * buffer, uint32_t buf_length, uint8_t command, uint8_t destination_endpoint)
+{
+    uint32_t result = 0;
+    // pick cluster id as 6 for now.
+    // pick source and destination end points as 1 for now.
+    // Profile is 65535 because that matches our simple generated code, but we
+    // should sort out the profile situation.
+
+    if (buffer == nullptr)
+    {
+        buf_length = 0; // This is the measuring case.
```

Comment:
This function shouldn't really have a measuring case per its api docs so far.  I'd rather we didn't do this until we do add one, because it would need more changes than this at that point.